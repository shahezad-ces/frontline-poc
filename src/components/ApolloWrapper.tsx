"use client";
import { ApolloProvider } from "@apollo/client/react";
import client from "@frontline/libs/apolloClient";

export default function ApolloWrapper({ children }: { children: React.ReactNode }) {
  return <ApolloProvider client={client}>{children}</ApolloProvider>;
}