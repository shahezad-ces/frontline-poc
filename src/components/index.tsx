import Categories from "./Categories/Categories";
import PaginationWrapper from "./PaginationWrapper/PaginationWrapper";
import ProductCard from "./ProductCard/ProductCard";
import ProductsFilters from "./ProductsFilters/ProductsFilters";
import AddToCartButton from "./AddToCartButton/AddToCartButton";
import Header from "./Header/Header";

export {
  Categories,
  PaginationWrapper,
  ProductCard,
  ProductsFilters,
  AddToCartButton,
  Header,
};
