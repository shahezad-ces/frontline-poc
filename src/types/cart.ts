import { Product } from "./product";

export type CartItem = Product & { quantity: number };

export type CartContextType = {
  cart: CartItem[];
  addToCart: (product: Product) => void;
  updateQuantity: (quantity: number) => void;
  removeFromCart: (id: number) => void;
  clearCart: () => void;
  totalItems: number;
  totalPrice: number;
  quantity: number;
};
