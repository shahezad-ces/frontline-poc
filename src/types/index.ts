import { Category } from "./category";
import { Product } from "./product";
import { CartItem, CartContextType } from "./cart";

export type { Category, Product, CartItem, CartContextType };
