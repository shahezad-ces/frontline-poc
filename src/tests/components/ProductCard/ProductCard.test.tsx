import { fireEvent, screen } from "@testing-library/react";
import { Product } from "@frontline/types";
import "@testing-library/jest-dom";
import { ProductCard } from "@frontline/components";
import { renderWithProviders } from "@frontline/tests/RenderWithProviders";

const baseProduct: Product = {
  id: 1,
  title: "Classic Red Hoodie",
  price: 49.99,
  images: ["https://placehold.co/600x400"],
  category: { name: "Clothing", id: 1, image: "test" },
  description: "A stylish hoodie for everyday wear.",
};

describe("ProductCard", () => {
  it("renders product details correctly", () => {
    renderWithProviders(<ProductCard {...baseProduct} />);
    expect(screen.getByText(/Classic Red Hoodie/i)).toBeInTheDocument();
    expect(screen.getByText(/\$49.99/i)).toBeInTheDocument();
    expect(screen.getByText(/Clothing/i)).toBeInTheDocument();
  });

  it("renders image with correct src and alt", () => {
    renderWithProviders(<ProductCard {...baseProduct} />);
    const img = screen.getByTestId("product-image");
    expect(img).toHaveAttribute("src", baseProduct.images[0]);
    expect(img).toHaveAttribute("alt", baseProduct.title);
  });

  it("renders AddToCartButton", () => {
    renderWithProviders(<ProductCard {...baseProduct} />);
    expect(screen.getByTestId("add-to-cart")).toHaveTextContent(`Add to Cart`);
  });

  it("links to correct product page", () => {
    renderWithProviders(<ProductCard {...baseProduct} />);
    expect(screen.getByRole("link")).toHaveAttribute(
      "href",
      `/product/${baseProduct.id}`
    );
  });

  it("handles price as 0", () => {
    const productWithZeroPrice = { ...baseProduct, price: 0 };
    renderWithProviders(<ProductCard {...productWithZeroPrice} />);
    expect(screen.getByText(/\$0/i)).toBeInTheDocument();
  });

  it("adds product to cart when Add to Cart button is clicked", () => {
    renderWithProviders(<ProductCard {...baseProduct} />);

    const addToCartButton = screen.getByTestId("add-to-cart");
    expect(addToCartButton).toBeInTheDocument();

    fireEvent.click(addToCartButton);
    expect(screen.getByText(/Item added to the cart/i)).toBeInTheDocument();
  });

  it("matches snapshot", () => {
    const { asFragment } = renderWithProviders(
      <ProductCard {...baseProduct} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
