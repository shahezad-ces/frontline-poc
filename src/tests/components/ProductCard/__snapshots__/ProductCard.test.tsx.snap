// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ProductCard matches snapshot 1`] = `
<DocumentFragment>
  <a
    href="/product/1"
  >
    <div
      class="bg-white shadow border border-gray-200 rounded-lg"
    >
      <img
        alt="Classic Red Hoodie"
        blurdataurl="/assets/images/placeholder.jpg"
        class="w-full h-40 md:h-48 lg:h-56 object-cover rounded-t-md"
        data-testid="product-image"
        height="500"
        placeholder="blur"
        src="https://placehold.co/600x400"
        width="500"
      />
      <div
        class="p-4"
      >
        <h2
          class="text-l font-semibold text-gray-800 line-clamp-2 h-15"
        >
          Classic Red Hoodie
        </h2>
        <p
          class="text-sm text-gray-400"
        >
          Clothing
        </p>
        <p
          class="text-xl text-red-600 font-semibold mt-1"
        >
          $49.99
        </p>
        <p
          class="text-sm text-black mt-1 line-clamp-2 mb-3 h-10"
        >
          Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
        </p>
        <button
          class="rounded-full font-semibold transition-colors duration-200 focus:outline-none bg-blue-800 text-white hover:bg-blue-700 px-5 py-2 text-sm disabled:bg-blue-300 disabled:text-gray-100 disabled:cursor-not-allowed w-full"
          data-testid="add-to-cart"
        >
          Add to Cart
        </button>
      </div>
    </div>
  </a>
</DocumentFragment>
`;
